torchvision/transforms/transforms.py

```diff
@@ -1953,3 +1953,126 @@ def forward(self, img):
 
     def __repr__(self):
         return self.__class__.__name__ + '(p={})'.format(self.p)
+
+
+class RandomMixupCutmix(torch.nn.Module):
+    """Randomly apply Mixum or Cutmix to the provided batch and targets.
+    The class implements the data augmentations as described in the papers
+    `"mixup: Beyond Empirical Risk Minimization" <https://arxiv.org/abs/1710.09412>`_ and
+    `"CutMix: Regularization Strategy to Train Strong Classifiers with Localizable Features"
+    <https://arxiv.org/abs/1905.04899>`_.
+
+    Args:
+        num_classes (int): number of classes used for one-hot encoding.
+        p (float): probability of the batch being transformed. Default value is 1.0.
+        mixup_alpha (float): hyperparameter of the Beta distribution used for mixup.
+            Set to 0.0 to turn off. Default value is 1.0.
+        cutmix_p (float): probability of using cutmix instead of mixup when both are on.
+            Default value is 0.5.
+        cutmix_alpha (float): hyperparameter of the Beta distribution used for cutmix.
+            Set to 0.0 to turn off. Default value is 0.0.
+        label_smoothing (float): the amount of smoothing using when one-hot encoding.
+            Set to 0.0 to turn off. Default value is 0.0.
+        inplace (bool): boolean to make this transform inplace. Default set to False.
+    """
+
+    def __init__(self, num_classes: int,
+                 p: float = 1.0, mixup_alpha: float = 1.0,
+                 cutmix_p: float = 0.5, cutmix_alpha: float = 0.0,
+                 label_smoothing: float = 0.0, inplace: bool = False) -> None:
+        super().__init__()
+        assert num_classes > 0, "Please provide a valid positive value for the num_classes."
+        assert mixup_alpha > 0 or cutmix_alpha > 0, "Both alpha params can't be zero."
+
+        self.num_classes = num_classes
+        self.p = p
+        self.mixup_alpha = mixup_alpha
+        self.cutmix_p = cutmix_p
+        self.cutmix_alpha = cutmix_alpha
+        self.label_smoothing = label_smoothing
+        self.inplace = inplace
+
+    def _smooth_one_hot(self, target: Tensor) -> Tensor:
+        N = target.shape[0]
+        device = target.device
+        v = torch.full(size=(N, 1), fill_value=1 - self.label_smoothing, device=device)
+        return torch.full(size=(N, self.num_classes), fill_value=self.label_smoothing / self.num_classes,
+                          device=device).scatter_add_(1, target.unsqueeze(1), v)
```

Comment:
Do we need to have this function be label-smoothing aware? I wonder if label smoothing in PyTorch can natively support this use-case https://github.com/pytorch/pytorch/pull/63122/files#diff-0b104a4612220d611b2c5d76272c28ddb21efbd8f9b9ae29d7f5b6c9e2a9be1fR480

Reply:
Yes we can. It's mathematically equivalent and it can help us decouple the two which is great.

From my original discussions with @jbschlosser, I understood that his plan was to create the `label_smoothing` argument only for the simple case where the inputs are labels and not probabilities. But then it seems after they changed plan on the implementations, they ended adding support also when the provided vectors are probabilities. This is great and simplifies things.